### ISSUE TITLE
tf.contrib.quantize bug: errors within weights/activations quantization

### ISSUE THREAD

**Issue by fgr1986** [tensorflow/tensorflow#20694]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes, provided
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: 1.8.0
- **Python version**:  3.6
- **Bazel version (if compiling from source)**: -
- **GCC/Compiler version (if compiling from source)**: - 
- **CUDA/cuDNN version**: -
- **GPU model and memory**: -
- **Exact command to reproduce**: python quantization_test.py (file provided)

### Describe the problem
**tf.contrib.quantize** (experimental and non-experimental) automatic quantization produce close to but not quantized results as when you manually introduce fakequant operations.
Weights and/or activations are close to the quantized values, but they are not correct.

### Source code / logs

```
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt

D_SIZE = 2048
BATCH_SIZE = 128
LR = 1e-2
EPOCHS = 2000

SHOW_INTERVAL = 20
OUT_CH = 8


def create_dummy_model(is_training=True):
    # prepare dataset
    np_features = np.arange(-int(D_SIZE) / 2, int(D_SIZE / 2))
    features_arr = tf.constant(np_features, dtype=tf.float32)
    np_labels = np.zeros((D_SIZE, OUT_CH), dtype=float)
    print(np_labels.shape)
    for i in range(OUT_CH):
        np_labels[:, i] = 2 * (i + 1) * np_features
    print('np_features:')
    print(np_features[0:8])
    print('np_labels:')
    print(np_labels[0:8])

    labels_arr = tf.constant(np_labels, dtype=tf.float32)
    trn_data = tf.data.Dataset.zip(
        (tf.data.Dataset.from_tensor_slices(features_arr),
         tf.data.Dataset.from_tensor_slices(labels_arr))
    )
    trn_data = trn_data.shuffle(D_SIZE)
    trn_data = trn_data.batch(BATCH_SIZE)
    trn_data = trn_data.repeat(1)
    trn_data = trn_data.prefetch(1)

    iterator = tf.data.Iterator.from_structure(trn_data.output_types,
                                               trn_data.output_shapes)
    features, labels = iterator.get_next()
    data_init_op = iterator.make_initializer(trn_data)

    # global step
    global_step = tf.train.get_or_create_global_step()

    # simple model dense, WITH ACTIVATION TO TEST QUANTIFICATION
    input = tf.reshape(features, (BATCH_SIZE, 1))
    flow = tf.identity(input, name='input_tensor')
    # batch_size x INPUT_CH
    flow = tf.reshape(flow, (-1, 1))
    flow = tf.layers.dense(flow, units=OUT_CH, name='dense')
    flow = tf.reshape(flow, (-1, OUT_CH), name='out')

    # Build forward pass of model.
    # Make the loss op
    with tf.variable_scope('loss'):
        tf.logging.debug('[loss] size of out: %s',
                         flow.get_shape().as_list())
        power = tf.pow(labels - flow, 2)
        power = tf.reduce_sum(power, 1)
        # power = tf.reduce_sum(power)
        loss = tf.reduce_mean(power, name='m_loss')
    tf.summary.scalar('loss', loss)

    # Optimizer
    if is_training:
        opt_op = tf.train.AdamOptimizer(LR).minimize(loss, global_step)
        # opt_op = tf.train.GradientDescentOptimizer(1e-4).minimize(loss, global_step)
    else:
        opt_op = False
    # Merge all the summaries
    merged = tf.summary.merge_all()

    return {
        'iterator': iterator,
        'features': features,
        'labels': labels,
        'data_init_op': data_init_op,
        'global_step': global_step,
        'flow': flow,
        'loss': loss,
        'opt_op': opt_op,
        'merged': merged
    }


def quantize_tensor(np_array, n_bits):
    """Quantizes an np_array"""
    min_wt = np.min(np_array)
    max_wt = np.max(np_array)
    # clap values if required
    if min_wt < -2**(n_bits-1):
        min_wt = -2**(n_bits-1)
    if max_wt > (2**(n_bits-1) - 1) :
        max_wt = (2**(n_bits-1) - 1)
    # find number of integer bits to represent this range
    int_bits = int(np.ceil(np.log2(max(abs(min_wt), abs(max_wt)))))
    print('int_bits: ', int_bits)
    # remaining bits are fractional bits (1-bit for sign)
    frac_bits = n_bits -1 - int_bits
    # floating point weights are scaled and rounded to [-128,127], which are used in
    # the fixed-point operations on the actual hardware (i.e., microcontroller)
    quantized = np.round(np_array * (2**frac_bits))
    # To quantify the impact of quantized weights, scale them back to
    # original range to run inference using quantized weights
    return quantized / (2**frac_bits)


def train_quantization(quantize=True, w_b=3, a_b=3):

    g_tr = tf.Graph()
    with g_tr.as_default():

        model = create_dummy_model(True)

        # Call the training rewrite which rewrites the graph in-place with
        # FakeQuantization nodes and folds batchnorm for training. It is
        # often needed to fine tune a floating point model for quantization
        # with this training tool. When training from scratch, quant_delay
        # can be used to activate quantization after training to converge
        # with the float graph, effectively fine-tuning the model.
        if quantize:
            quant_delay = int((D_SIZE / BATCH_SIZE) * EPOCHS * 3 / 4)
            print('quant_delay: ', quant_delay)
            print('w_b: ', w_b)
            print('a_b: ', a_b)
            tf.contrib.quantize.experimental_create_training_graph(weight_bits=w_b,
                                                                   activation_bits=a_b,
                                                                   quant_delay=quant_delay)

        with tf.Session(graph=g_tr) as sess:
            # global variables Initializing
            sess.run(tf.global_variables_initializer())
            # local variables Initializing
            sess.run(tf.local_variables_initializer())

            if quantize:
                train_writer = tf.summary.FileWriter(
                    './tmp/train_quantized_w' + str(w_b) + '_a_' + str(a_b),
                    sess.graph)
            else:
                train_writer = tf.summary.FileWriter('./tmp/train_normal',
                                                     sess.graph)

            epoch_counter = 0
            nn_in = None
            nn_out = None
            nn_loss = None
            for epoch_counter in range(EPOCHS):

                # re-initialize the dataset iterator
                sess.run(model['data_init_op'])

                try:
                    while True:
                        _, nn_in, nn_out, nn_labels, nn_loss, gs, summary = sess.run(
                            [model['opt_op'],
                             model['features'],
                             model['flow'],
                             model['labels'],
                             model['loss'],
                             model['global_step'],
                             model['merged']])

                        # tensorboard and  statistics
                        train_writer.add_summary(summary, gs)
                except tf.errors.OutOfRangeError:
                    # do nothing
                    if epoch_counter % SHOW_INTERVAL == 0:
                        print(epoch_counter,  '/', EPOCHS,
                              ': loss ', nn_loss,
                              '. Global step: ', gs)

            weights = sess.run(tf.get_default_graph().get_tensor_by_name(
                'dense/kernel:0'))
            saver = tf.train.Saver()
            if quantize:
                trained_model_path = './tmp/quantized_w' + \
                    str(w_b) + '_a_' + str(a_b) + '.ckpt'
            else:
                trained_model_path = './tmp/normal.ckpt'

            saver.save(sess, trained_model_path)

        # debug
        print('nn_out shape: ', nn_out.shape)
        print('nn_labels shape: ', nn_labels.shape)
        print('nn_out different values in OUT_CH (', OUT_CH, ') elements:',
              len(np.unique(nn_out)))
        print('weights shape: ', weights.shape)
        print('weights: ', weights)
        print('weights different values in OUT_CH (', OUT_CH, ') elements:',
              len(np.unique(weights)))

        plt.plot(nn_in[:], nn_out[:, 0], label='nn_out',
                 marker='o', markersize=5)
        plt.plot(nn_in[:], nn_labels[:, 0], label='labels',
                 marker='*', markersize=5)
        plt.legend(loc='lower right')
        plt.show()
        plt.plot(nn_in[:], nn_out[:, 5], label='nn_out',
                 marker='o', markersize=5)
        plt.plot(nn_in[:], nn_labels[:, 5], label='labels',
                 marker='*', markersize=5)
        plt.legend(loc='lower right')
        plt.show()

        plt.plot(weights[0, :], label='weights',
                 marker='o', markersize=5)
        plt.legend(loc='lower right')
        plt.show()

        print('[debug] exit training')
        return trained_model_path


def inference_quantization(trained_model_path, quantize=True, w_b=3, a_b=3):

    #################################
    # Reset default graph
    tf.reset_default_graph()
    #################################

    g_inf = tf.Graph()
    with g_inf.as_default():

        model = create_dummy_model(False)

        with tf.Session(graph=g_inf) as sess:
            # Call the eval rewrite which rewrites the graph in-place with
            # FakeQuantization nodes and fold batchnorm for eval.
            if quantize:
                print('w_b: ', w_b)
                print('a_b: ', a_b)
                tf.contrib.quantize.experimental_create_eval_graph(
                    weight_bits=w_b,
                    activation_bits=a_b)

            saver = tf.train.Saver()
            saver.restore(sess, trained_model_path)
            # frozen_graphdef = tf.graph_util.convert_variables_to_constants(
            #       sess, sess.graph_def, ["out"])
            print('[debug] restored')

            # re-initialize the iterator, but this time with training data

            sess.run(model['data_init_op'])

            nn_in, nn_out, nn_labels, nn_loss = sess.run([
                model['features'],
                model['flow'],
                model['labels'],
                model['loss']])
            weights = sess.run(tf.get_default_graph(
            ).get_tensor_by_name('dense/kernel:0'))

        # debug
        print('nn_out shape: ', nn_out.shape)
        print('nn_labels shape: ', nn_labels.shape)
        print('nn_out different values in OUT_CH (', OUT_CH, ') elements:',
              len(np.unique(nn_out)))
        print('weights shape: ', weights.shape)
        # print('weights: ', weights)
        print('weights different values in OUT_CH (', OUT_CH, ') elements:',
              len(np.unique(weights)))
        print('weights: ', weights)
        q_weights = quantize_tensor(weights, w_b)
        print('q_weights: ', q_weights)
        plt.plot(nn_in[:], nn_out[:, 0], label='nn_out',
                 marker='o', markersize=5)
        plt.plot(nn_in[:], nn_labels[:, 0], label='labels',
                 marker='*', markersize=5)
        plt.legend(loc='lower right')
        plt.show()
        plt.plot(nn_in[:], nn_out[:, 5], label='nn_out',
                 marker='o', markersize=5)
        plt.plot(nn_in[:], nn_labels[:, 5], label='labels',
                 marker='*', markersize=5)
        plt.legend(loc='lower right')
        plt.show()

        plt.plot(weights[0, :], label='weights',
                 marker='o', markersize=5)
        plt.legend(loc='lower right')
        plt.show()


def test_quantization():

    # train_quantization(False)
    trained_model_path = train_quantization(quantize=True, w_b=2, a_b=2)
    inference_quantization(
        trained_model_path=trained_model_path, quantize=True, w_b=2, a_b=2)

test_quantization()
```

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @poxvoculi: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by suharshs** [comment]
This is because the FakeQuant operations do more than just quantize. First its very important that floating point 0.0 maps directly to a quantized value, otherwise padding errors can propagate. This nudging can be seen here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/fake_quant_ops_functor.h#L41

Also we restrict the weight values to being quantized from [1,255] instead of [0,255] to enable a ARM Neon optimization for our kernels.

Thus, this is working as intended, please do reopen and correct me if I am misunderstanding or mistaken.

**Comment by suharshs** [comment]
Are your sure you are running the FakeQuant nodes after the weights. The weights will still retain their full precision but the output fo the following FakeQuant will emulate the correct quantization. So make sure for each weight you are actually checking the 'weight_quant/FakeQuantWithMinMaxVars' operation following it.

**Comment by fgr1986** [comment]
You were completely right. only the Bias nodes were not quantized.
Thanks.
